tests/unit-tests/admin/home/test-class-sensei-home-remote-data-api.php

```diff
@@ -0,0 +1,255 @@
+		$remote_data   = $provider->fetch();
+
+		$this->stopTrackingHttpRequests();
+		remove_all_filters( 'sensei_home_remote_data_primary_plugin' );
+
+		$this->assertEquals( 1, $http_requests->count() );
+		$expected_url   = Sensei_Home_Remote_Data_API::API_BASE_URL . 'penguins.json';
+		$requested_path = strtok( $http_requests[0]['url'], '?' );
+		$this->assertEquals( $expected_url, $requested_path );
+	}
+
+	/**
+	 * Tests to make sure the cache is used when hydrated.
+	 */
+	public function testFetchCacheUsed() {
+		$http_requests = $this->trackHttpRequests(
+			function() {
+				return [
+					'hit' => uniqid(),
+				];
+			}
+		);
+
+		$provider     = new Sensei_Home_Remote_Data_API( 'dinosaurs' );
+		$first_fetch  = $provider->fetch();
+		$second_fetch = $provider->fetch();
+		$this->stopTrackingHttpRequests();
+
+		$this->assertIsArray( $first_fetch );
+		$this->assertIsArray( $second_fetch );
+
+		$this->assertEquals( $first_fetch['hit'], $second_fetch['hit'], 'The two requests should return the same unique ID' );
+	}
+
+	/**
+	 * Tests to make sure the max age is respected.
+	 */
+	public function testFetchCacheMaxAgeRespected() {
+		$http_requests = $this->trackHttpRequests(
+			function() {
+				return [
+					'hit' => uniqid(),
+				];
+			}
+		);
+
+		// Test max age of 60 seconds.
+		$max_age = 60;
+		$url     = Sensei_Home_Remote_Data_API::API_BASE_URL . '/test.json';
+
+		$provider = $this->getMockBuilder( Sensei_Home_Remote_Data_API::class )
+			->setConstructorArgs( [ 'dinosaurs' ] )
+			->setMethods( [ 'get_api_url' ] )
+			->getMock();
+		$provider->expects( $this->any() )->method( 'get_api_url' )->willReturn( $url );
+
+		// Clone to avoid local caches.
+		$provider_b = clone $provider;
+
+		$first_fetch = $provider->fetch( $max_age );
+
+		// Artificially change the fetched time of the cached data.
+		$this->artificiallyChangeCache( $url, [ '_fetched' => time() - $max_age - 1 ] );
+
+		$second_fetch = $provider_b->fetch( $max_age );
+		$this->stopTrackingHttpRequests();
+
+		$this->assertIsArray( $first_fetch );
+		$this->assertIsArray( $second_fetch );
+
+		$this->assertNotEquals( $first_fetch['hit'], $second_fetch['hit'] );
+	}
+
+	/**
+	 * Tests to make sure local cache is used when hydrated.
+	 */
+	public function testFetchLocalCacheUsed() {
+		$http_requests = $this->trackHttpRequests(
+			function() {
+				return [
+					'hit' => uniqid(),
+				];
+			}
+		);
+
+		// Test max age of 60 seconds.
+		$max_age = 60;
+		$url     = Sensei_Home_Remote_Data_API::API_BASE_URL . '/test.json';
+
+		$provider = $this->getMockBuilder( Sensei_Home_Remote_Data_API::class )
+			->setConstructorArgs( [ 'dinosaurs' ] )
+			->setMethods( [ 'get_api_url' ] )
+			->getMock();
+		$provider->expects( $this->any() )->method( 'get_api_url' )->willReturn( $url );
+
+		$first_fetch = $provider->fetch( $max_age );
+
+		// Artificially change the fetched time of the cached data.
+		$this->artificiallyChangeCache( $url, [ '_bad' => true ] );
```

Comment:
Wrong comment. Not really changing fetched time.

Reply:
Fixed in 8a9e2b7fe7774588eb39ba3af38f9cb9be4f5dd2